### ISSUE TITLE
S3 object storage

### ISSUE THREAD

**Issue by noplanman** [friendica/friendica#8457]
Are there any plans to implement S3 object storage for all media?
Or is there maybe already an undocumented way to do this?

**Comment by annando** [comment]
Currently there are no plans.

**Comment by MrPetovan** [comment]
@fabrixxm created the storage adapters, maybe he has a few pointers?

**Comment by fabrixxm** [comment]
Well, as storage backends can be implemented as addons, the starting point is certainly  the official docs: https://social.gl-como.it/help/AddonStorageBackend

Is to be noted that for how friendica works, one could use S3 to store and retrieve media data (having a backend), but can't be used to distribute media to end users: everything has to pass thru friendica code. 
(e.g. you can't have photos on a cdn)

**Comment by noplanman** [comment]
Hi @fabrixxm, thanks for the reply.

Would you mind explaining the "everything has to pass through friendica code" a bit more?

So uploading to S3 instead of local storage and rewriting the media URLs to the S3 bucket won't work?

**Comment by AlfredSK** [comment]
I might be wrong but that would completely circumvent the permission system.

**Comment by fabrixxm** [comment]
No, friendica has the concept of private media: you can decide which contacts can see your media.
Every request to media must be checked for permissions. Storage backends are build around this idea. They only manage where and how store data, but the frontend url remain the same always.

This is to avoid the "facebook non-private photos" situation, where a private photo has a public, non authenticated URL.

Note that also the default Filesystem Storage works this way. Photo are never served directly by the webserver from the filesystem.

**Comment by noplanman** [comment]
Right, I understand what you mean now, thanks so much for the explanation!

So basically the media can't be served by a reverse proxy listening on the media files path, as that would create the same situation :thinking: 

In that case any S3 (or any other external) storage will add a delay to media loading times.

Will need to look into the docs and code to understand how the flow of storage data works exactly.

Thanks again :pray:

**Comment by hoergen** [comment]
Maybe a way to use it, is to mount (operating system wise) that S3 storage into your local filesystem?

**Comment by noplanman** [comment]
@hoergen Yes, that does seem to be the only way for now. Will explore that further, thanks.
